### ISSUE TITLE
load_model not properly working in TF 2.3.0

### ISSUE THREAD

**Issue by HaraBeron** [tensorflow/tensorflow#42890]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Custom Code
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows
- TensorFlow installed from (source or binary): Colab TF 2.3.0 vs TF 2.2.0
- TensorFlow version (use command below): Colab TF 2.3.0 vs TF 2.2.0
- Python version: Colab Python
- GPU model and memory: Colab GPU

When I used TF 2.2.0 in colab I was able to fit, save and load a model. So I run my code, fit my model:
![grafik](https://user-images.githubusercontent.com/68148852/91966779-9bf5e880-ed12-11ea-8ec3-7e499a5024eb.png)

 and check again with model.evaluate:
![grafik](https://user-images.githubusercontent.com/68148852/91966830-ac0dc800-ed12-11ea-9ab8-68c2b84cb218.png)

Everything as expected. Then I save the model and load it back again and I check it again with model.evaluate and same accuracy is shown. Everything fine. However, the same code does not work in TF 2.3.0. See my code below. If this code is run by default TF 2.3.0 is used in colab and it does not work. The accuracy of the loaded model loadedmodel.evaluate(x_test_r,test_labels) is not the same:

![grafik](https://user-images.githubusercontent.com/68148852/91966880-c051c500-ed12-11ea-9b72-3c57fffc5646.png)

It is completely wrong (approx .098... vs supposed 0.99... when I run it). When I add !pip install tensorflow==2.2.0 to my code and run the same code in my colab again (note that the runtime has to be restarted) then with the 2.2.0 version it does work.

(When I save the model in TF 2.2.0 (where evaluate works and correct accuracy is shown) and then load it in TF 2.3.0, it does not work. model.evaluate shows a wrong accuracy.)

**Standalone code to reproduce the issue**
```
#!pip install tensorflow==2.2.0
import tensorflow as tf
import os
import PIL
import numpy as np
import matplotlib.pyplot as plt
import tensorflow_datasets as tfds
import tensorflow_hub as hub

from tensorflow.keras.models import Sequential

from tensorflow.keras.layers import Dense, Conv2D, Flatten, Dropout, MaxPooling2D
from tensorflow.keras.preprocessing.image import ImageDataGenerator

import keras_preprocessing
from keras_preprocessing import image

from tensorflow.python.keras.utils.version_utils import training
from tensorflow.keras.optimizers import RMSprop

print(tf.__version__)

(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.mnist.load_data()

x_trainf = train_images.astype('float32') / 255.0
x_testf = test_images.astype('float32') / 255.0

x_train_r = x_trainf.reshape(x_trainf.shape[0], 28, 28, 1)
x_test_r = x_testf.reshape(x_testf.shape[0], 28, 28, 1)

model = tf.keras.Sequential()
model.add(tf.keras.layers.Conv2D(filters=64, kernel_size=3, padding='same', activation='relu', input_shape=(28,28,1))) 
model.add(tf.keras.layers.MaxPooling2D(pool_size=2))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Conv2D(filters=32, kernel_size=2, padding='same', activation='relu'))
model.add(tf.keras.layers.MaxPooling2D(pool_size=2))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(256, activation='relu'))
model.add(tf.keras.layers.Dropout(0.2))
model.add(tf.keras.layers.Dense(10))

model.compile(optimizer='adam',
              loss=tf.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(x_train_r, 
          train_labels, batch_size=32,  
          epochs=10,
          validation_data=(x_test_r,test_labels))

model.evaluate(x_test_r,test_labels)

model.save('/tmp/loaddatamnist.h5')

loadedmodel=tf.keras.models.load_model('/tmp/loaddatamnist.h5')
loadedmodel.evaluate(x_test_r,test_labels)
```

**Comment by ravikyram** [comment]
@HaraBeron 

I have tried in colab with TF nightly version(2.4.0-dev20200902) and i am not seeing any issue. Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/4cb33851aa5e54ab42b787f9734773f2/untitled297.ipynb).Please, verify once and close the issue. Thanks!

**Comment by HaraBeron** [comment]
I was not talking about TF nightly version 2.4.0-... I was talking about TF 2.2.0 and TF 2.3.0 and there the issue exists and this is what most users are using. They do not use nightly version.

When you run your colab without the first line, so without !pip install tf-nightly. Then TF 2.3.0 is used and then the same issue occurs. I run your code without this line and the issue as described occurs.

In detail:
When I run your code (with just 2 epochs) I get an val_accuracy of 0.9881 when I use model.evaluate it gives the same 0.988099. Then I save and load the model and run model.evaluate again. This should give the same value, however it doesn't. It gives a wrong value of 0.099.

So the issue does exist.

**Comment by HaraBeron** [comment]
Here is a screenshot of your colab when I run it:

![grafik](https://user-images.githubusercontent.com/68148852/91978231-13804380-ed24-11ea-8718-c5863a101cec.png)

There you can see the issue (here in this case it waas not 0.099, but 0.1, still a wrong value). It is without the tf nightly. It is TF 2.3.0.

When you add !pip install tensorflow==2.2.0 instead of the tf nightly and then run it (you maybe have to restart the runtime), you can see, that it does work. Here is a screenshot:

![grafik](https://user-images.githubusercontent.com/68148852/91978716-e54f3380-ed24-11ea-96e9-97ba9d3e5519.png)

So again: It works in TF 2.2.0, but not in TF 2.3.0.

**Comment by ravikyram** [comment]
@HaraBeron 

I already tried in TF 2.3 and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/ea2d084ee5d54411a0dbb545b13ca7e6/untitled295.ipynb).

**Comment by bhack** [comment]
I don't know the compiled status it is recovered. Please try to recompile after reloading:

```
loadedmodel=tf.keras.models.load_model('/tmp/test.h5')
loadedmodel.compile(optimizer='adam',
              loss=tf.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
loadedmodel.evaluate(x_test_r,test_labels)
```

**Comment by jvishnuvardhan** [comment]
@HaraBeron This is a know bug with `accuracy`. In earlier TF versions `accuracy` was correctly inferred from the `loss` function. In this case, `accuracy` need to be inferred as `sparse_categorical_accuracy` but in the current version it is not working as expected.

Everything works as expected If you replace `accuracy` with `sparse_categorical_accuracy` in `model.compile` as shown below.

```
model.compile(optimizer='adam',
              loss=tf.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['sparse_categorical_accuracy'])
```

Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/933388973689f871e75248e022e33800/untitled295.ipynb). 

We will track the progress of this bug in [this issue](https://github.com/tensorflow/tensorflow/issues/42045).

Please verify once and close the issue. We will track the progress of this bug in the above mentioned issue. 

I will also comment when everything is working as intended. Thanks!

**Comment by jvishnuvardhan** [comment]
@HaraBeron This is already fixed in recent `tf-nightly`. This will be available in stable `TF2.4` version. Please take a look at the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/3311c392b3a3112fdd1e2ac4f7e5c580/untitled295.ipynb). Thanks!

**Comment by jvishnuvardhan** [comment]
This is resolved in recent tf-nightly. Please feel free to reopen if you notice the issue.
This is available in stable TF2.4 in the near future. Thanks!
